modules/calib3d/src/ap3p.cpp

```diff
@@ -256,6 +256,8 @@ int ap3p::computePoses(const double featureVectors[3][4],
     int nb_solutions = 0;
     for (int i = 0; i < 4; ++i) {
         double ctheta1p = s[i];
+        if (isnan(ctheta1p))
+            continue;
```

Comment:
Hi! Can you provide a reproducer so this can be covered by some test?

Reply:
Hi!
Sorry for the delay. Here is the reproducer:

```
{
    const float fx = 1, fy = 1, cx = 0, cy = 0;
    cv::ap3p solver(fx, fy, cx, cy);

    const std::array<cv::Point2d, 3> cameraPts = {                //
        cv::Point2d{0.042784865945577621, 0.59844839572906494},   //
        cv::Point2d{-0.028428621590137482, 0.60354739427566528},  //
        cv::Point2d{0.0046037044376134872, 0.70674681663513184}   //
    };

    const std::array<cv::Point3d, 3> modelPts = {                                          //
        cv::Point3d{-0.043258000165224075, 0.020459245890378952, -0.0069921980611979961},  //
        cv::Point3d{-0.045648999512195587, 0.0029820732306689024, 0.0079000638797879219},  //
        cv::Point3d{-0.043276999145746231, -0.013622495345771313, 0.0080113131552934647}   //
    };

    double R[4][3][3];
    double t[4][3];
    const int nb_solutions = solver.solve(R, t,                                       //
        cameraPts[0].x, cameraPts[0].y, modelPts[0].x, modelPts[0].y, modelPts[0].z,  //
        cameraPts[1].x, cameraPts[1].y, modelPts[1].x, modelPts[1].y, modelPts[1].z,  //
        cameraPts[2].x, cameraPts[2].y, modelPts[2].x, modelPts[2].y, modelPts[2].z,  //
        0, 0, 0, 0, 0, false);
} 
```

Reply:
@alexander-varjo, Cannot reproduce, the values of `s` array are the following:
```
7.5499e+06
7.5499e+06
-1.6255e+07
-1.6255e+07
```

Reply:
Please share compiler version and target architecture. On GCC 9.4.0 and Intel Core i5-1135G7 the values of `s` array are not NaN:
```
s[0]: 7.5499e+06
s[1]: 7.5499e+06
s[2]: -1.6255e+07
s[3]: -1.6255e+07
```

Reply:
My setup:
* Compiler: MSVC 14.29.30133
* CPU: i9-10850K

Did you run directly cv::ap3p::solve() or you just used cv::solveP3P() when you tested the reproducer?

Reply:
Both, for me problem is not reproduced so it might depend on a compiler

Reply:
Can reproduce with the following code on `MSVC 14.35.32215`. The app not crashed but nan returned.

Let me check if there is no other solution to this problem.

```patch
--- a/modules/calib3d/src/ap3p.cpp
+++ b/modules/calib3d/src/ap3p.cpp
@@ -3,6 +3,7 @@
 
 #include <cmath>
 #include <complex>
+#include <iostream>
 #if defined (_MSC_VER) && (_MSC_VER <= 1700)
 static inline double cbrt(double x) { return (double)cv::cubeRoot((float)x); };
 #endif
@@ -256,6 +257,7 @@ int ap3p::computePoses(const double featureVectors[3][4],
     int nb_solutions = 0;
     for (int i = 0; i < 4; ++i) {
         double ctheta1p = s[i];
+        std::cout << i << " " << ctheta1p << std::endl;
         if (abs(ctheta1p) > 1)
             continue;
         double stheta1p = sqrt(1 - ctheta1p * ctheta1p);
```
```cpp
#include <opencv2/opencv.hpp>

using namespace cv;

int main(int argc, char** argv) {
    const std::array<cv::Point2d, 3> cameraPts = {                //
        cv::Point2d{0.042784865945577621, 0.59844839572906494},   //
        cv::Point2d{-0.028428621590137482, 0.60354739427566528},  //
        cv::Point2d{0.0046037044376134872, 0.70674681663513184}   //
    };

    const std::array<cv::Point3d, 3> modelPts = {                                          //
        cv::Point3d{-0.043258000165224075, 0.020459245890378952, -0.0069921980611979961},  //
        cv::Point3d{-0.045648999512195587, 0.0029820732306689024, 0.0079000638797879219},  //
        cv::Point3d{-0.043276999145746231, -0.013622495345771313, 0.0080113131552934647}   //
    };
    std::vector<Mat> R, t;
    solveP3P(modelPts, cameraPts, Mat::eye(3, 3, CV_64F), Mat(), R, t, SOLVEPNP_AP3P);
    for (int i = 0; i < R.size(); ++i) {
        std::cout << R[i] << std::endl;
        std::cout << t[i] << std::endl;
        std::cout << " " << std::endl;
    }
    return 0;
}
```
```
0 nan
1 nan
2 -nan(ind)
3 -nan(ind)
```

Reply:
The issue is reproducible with ENABLE_FAST_MATH option in Cmake. I do not get crash, but R and t contains NaNs.